controllers/cache_controller.go

```diff
@@ -40,144 +61,397 @@ func (r *CacheReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		r.reqLogger.Error(err, "Error updating cache")
+		return err
+	}
+
+	if spec.TemplateName != "" || spec.Template != "" {
+		err := fmt.Errorf("cannot create a cache with a template in a CacheService cluster")
+		r.reqLogger.Error(err, "Error creating cache")
+		return err
+	}
+
+	template, err := caches.DefaultCacheTemplateXML(podName, r.infinispan, cluster, r.reqLogger)
+	if err != nil {
+		err = fmt.Errorf("unable to obtain default cache template: %w", err)
+		r.reqLogger.Error(err, "Error getting default XML")
+		return err
+	}
+
+	err = cluster.CreateCacheWithConfiguration(r.cache.Spec.Name, template, podName, mime.ApplicationXml)
+	if err != nil {
+		err = fmt.Errorf("unable to create cache using default template: %w", err)
+		r.reqLogger.Error(err, "Error in creating cache")
+		return err
+	}
+	return nil
+}
+
+func (r *cacheRequest) reconcileDataGrid(cacheExists bool, podName string, cluster infinispan.ClusterInterface) error {
+	spec := r.cache.Spec
+	cacheName := r.cache.GetCacheName()
+	if cacheExists {
+		if spec.TemplateName != "" {
+			// TODO enforce with webhook validation when supported
+			r.log.Error(fmt.Errorf("updating an existing Cache's 'spec.TemplateName' field is not supported"), "")
 		} else {
-			reqLogger.Info(fmt.Sprintf("Cache %s doesn't exist, create it", instance.GetCacheName()))
-			podName := podList.Items[0].Name
-			templateName := instance.Spec.TemplateName
-			if ispnInstance.Spec.Service.Type == infinispanv1.ServiceTypeCache && (templateName != "" || instance.Spec.Template != "") {
-				errTemplate := fmt.Errorf("cannot create a cache with a template in a CacheService cluster")
-				reqLogger.Error(errTemplate, "Error creating cache")
-				return reconcile.Result{}, err
+			err := cluster.UpdateCacheWithConfiguration(cacheName, spec.Template, podName, mime.GuessMarkup(spec.Template))
+			if err != nil {
+				return fmt.Errorf("unable to update cache template: %w", err)
 			}
-			if templateName != "" {
-				err = cluster.CreateCacheWithTemplateName(instance.Spec.Name, templateName, podName)
-				if err != nil {
-					reqLogger.Error(err, "Error creating cache with template name")
-					return reconcile.Result{}, err
+		}
+		return nil
+	}
+
+	var err error
+	if spec.TemplateName != "" {
+		if err = cluster.CreateCacheWithTemplateName(cacheName, spec.TemplateName, podName); err != nil {
+			err = fmt.Errorf("unable to create cache with template name '%s': %w", spec.TemplateName, err)
+		}
+	} else {
+		if err = cluster.CreateCacheWithConfiguration(cacheName, spec.Template, podName, mime.GuessMarkup(spec.Template)); err != nil {
+			err = fmt.Errorf("unable to create cache with template: %w", err)
+		}
+	}
+
+	if err != nil {
+		r.reqLogger.Error(err, "Unable to create Cache")
+	}
+	return err
+}
+
+func (cl *CacheListener) CreateOrUpdate(data []byte) error {
+	cacheName, configYaml, err := unmarshallEventConfig(data)
+	if err != nil {
+		return err
+	}
+
+	if strings.HasPrefix(cacheName, "___") {
```

Comment:
Not all internal caches have `___` as a prefix. There are also `org.infinispan.` caches.

Reply:
I wasn't sure if we want to include those or not ... I can update the check to include them. Shouldn't the server exclude all of these caches in the first place?

Reply:
Actually, the `ContainerResource#ConfigurationListener` returns the result of `EmbeddedCacheManager#getCacheConfigurationNames` for `includeCurrentState` which filters internal caches, only returning `___script_cache` and `___protobuf_metadata` internal caches.

@tristantarrant Is it possible for `ConfigurationChangedEvent` to be fired for any of the other internal caches?